File: src/mlf/expression.ts
import {Reported} from './diagnostics';

export type Constant =
  | {readonly kind: 'Boolean'; readonly value: boolean}
  | {readonly kind: 'Number'; readonly value: number}
  | {readonly kind: 'String'; readonly value: string};

export type Expression<Diagnostic = never, Type = undefined> = {
  readonly type: Type;
  readonly description: ExpressionDescription<Diagnostic, Type>;
};

export type ExpressionDescription<Diagnostic, Type> =
  | {
      readonly kind: 'Variable';
      readonly identifier: string;
    }
  | {
      readonly kind: 'Constant';
      readonly constant: Constant;
    }
  | {
      readonly kind: 'Function';
      readonly parameter: string;
      readonly body: Expression<Diagnostic, Type>;
    }
  | {
      readonly kind: 'Application';
      readonly callee: Expression<Diagnostic, Type>;
      readonly argument: Expression<Diagnostic, Type>;
    }
  | {
      readonly kind: 'Binding';
      readonly binding: string;
      readonly value: Expression<Diagnostic, Type>;
      readonly body: Expression<Diagnostic, Type>;
    }
  | {
      readonly kind: 'Error';
      readonly error: Reported<Diagnostic>;
    };

File: src/Expression.ml
type expression = {
  description: expression_description;
}

and expression_description =
  (* `x` *)
  | Variable of { name: string }

  (* `true`, `false` *)
  | Boolean of bool

  (* `0`, `42`, `3.1415`, `-1` *)
  | Number of float

  (* `""`, `"foo"` *)
  | String of string

  (* `λx.E` *)
  | Function of { parameter: string; body: expression }

  (* `E1 E2` *)
  | Call of { callee: expression; argument: expression }

  (* `let x = E1 in E2` *)
  | Binding of { name: string; value: expression; body: expression }

  (* `(E: T)` *)
  | Annotation of { value: expression; type_: Type.polytype }

type t = expression

let variable name =
  { description = Variable { name } }

let boolean value =
  { description = Boolean value }

let number value =
  { description = Number value }

let string value =
  { description = String value }

let function_ parameter body =
  { description = Function { parameter; body } }

let call callee argument =
  { description = Call { callee; argument } }

let binding name value body =
  { description = Binding { name; value; body } }

let annotation value type_ =
  { description = Annotation { value; type_ } }
